controller/source/logging/logger.hpp

```diff
@@ -0,0 +1,61 @@
+#pragma once
+#include <iostream>
+#include <fstream>
+#include <mutex>
+#include <unordered_map>
+#include <filesystem>
+
+namespace controller {
+
+/**
+ * Singleton logger class to store the history of each pair in a different file.
+*/
+class logger {
+public:
+  static auto get_instance() -> logger* {
+    static logger history_logger;
+    return &history_logger;
+  }
+
+  void log(const query& query_args, const bool& result, const std::string& message = {}) {
+    
+    // lock the mutex corresponding to the key
+    std::lock_guard<std::mutex> lock(m_keys_to_mutexes[query_args.key()]);
```

Comment:
With the timestamp that you include in the message, you can simply count the bytes, maintain an offset variable and perform an [atomic increment](https://en.cppreference.com/w/cpp/atomic/atomic/operator_arith) to point to the next position available for the next entry in the log.
It requires some more thought though.

Reply:
As agreed, let's keep things simple for the time being.